### PR TITLE
refactor!: separated transport crate

### PR DESCRIPTION
**Pull request by Ma233**
### Please check if the PR fulfills these requirements

- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes / features)

## :large_blue_circle: What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
This PR separated the whole transport module of rings-core crate into an individual crate.
The new rings-transport crate provided:
* `Transport` to hold all connections.
* `Connection` work as the old `Transport` objects.
* `Callback` trait and `BoxedCallback` trait object to help invoker pass callbacks to the `Connection`.

## :brown_circle: What is the current behavior? (You can also link to an open issue here)
The maintenance of Transport and the behavior of rings network are both managed in swarm.

## :green_circle: What is the new behavior (if this is a feature change)?
It allows the swarm to focus solely on network behavior (which may be divided into multiple protocols in the future 😄).
Supposed to have no new feature since this is a refactor.

## :radioactive: Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)
1. The create_offer API has been updated. Users now need to provide the peer's DID at the beginning of the handshake.
2. All APIs related to pending transport have been removed. Each connection is now unique to a specific Did. If you attempt to create a new connection during a handshake, you will encounter a ConnectionAlreadyExists error. You should either disconnect that connection or wait for it to be connected.

## :information_source: Other information

Closes #issue

### CODE REVIEW

File: core/src/message/types.rs [hunk 1]
```diff
@@ -22,19 +21,15 @@ pub trait Then {
 /// MessageType use to ask for connection, send to remote with transport_uuid and handshake_info.
 #[derive(Debug, PartialEq, Eq, Deserialize, Serialize, Clone)]
 pub struct ConnectNodeSend {
-    /// uuid of transport
-    pub transport_uuid: String,
     /// sdp offer of webrtc
-    pub offer: HandshakeInfo,
+    pub sdp: String,
```
Comment (RyanKung) [inline core/src/message/types.rs:25]: Why erasure type there?

File: core/src/swarm/builder.rs [hunk 2]
```diff
@@ -2,26 +2,26 @@
...
+use crate::swarm::callback::SwarmCallback;
 use crate::swarm::MeasureImpl;
 use crate::swarm::Swarm;
 use crate::types::channel::Channel as ChannelTrait;
-use crate::types::ice_transport::IceServer;
 
 /// Creates a SwarmBuilder to configure a Swarm.
 pub struct SwarmBuilder {
-    ice_servers: Vec<IceServer>,
+    ice_servers: String,
```
Comment (RyanKung) [inline core/src/swarm/builder.rs:24]: Also here, why erasure type here

File: core/src/swarm/types.rs [hunk 3]
```diff
@@ -21,27 +20,28 @@ pub type MeasureImpl = Box<dyn BehaviourJudgement + Send + Sync>;
...
 pub type MeasureImpl = Box<dyn BehaviourJudgement>;
 
-/// WrappedDid is a DID wrapped by Swarm and bound to a weak reference of a Transport,
+/// WrappedDid is a DID wrapped by Swarm and bound to a Connection,
 /// which enables checking whether the WrappedDid is live or not.
 #[derive(Clone)]
 pub struct WrappedDid {
     did: Did,
-    transport: Option<Weak<Transport>>,
+    connection: Option<Connection>,
```
Comment (RyanKung) [inline core/src/swarm/types.rs:28]: I think WrappedDid should not take the ownership of Connection. 
Once a bounded connection dropped, WrappedDid.live should return `false`.
Or, should it be bounded to transport, if transport is presenting for connection layer or management?

File: node/src/browser/client.rs [hunk 4]
```diff
@@ -310,20 +300,17 @@ impl Client {
...
-                peer.transport.id,
-            )))?)
+            let state = peer.connection.webrtc_connection_state();
+            Ok(JsValue::try_from(&Peer::from((state, peer.did)))?)
         })
     }
 
     /// Manually make handshake with remote peer
-    pub fn create_offer(&self) -> js_sys::Promise {
+    pub fn create_offer(&self, address: String, addr_type: Option<AddressType>) -> js_sys::Promise {
```
Comment (RyanKung) [inline node/src/browser/client.rs:309]: AddressType?

File: transport/src/callback.rs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+use bytes::Bytes;
+
+use crate::core::callback::BoxedCallback;
+use crate::core::transport::TransportMessage;
+use crate::core::transport::WebrtcConnectionState;
+use crate::notifier::Notifier;
+
+pub(crate) struct InnerCallback {
+    pub cid: String,
```
Comment (RyanKung) [inline transport/src/callback.rs:11]: what is cid for? need document here.

File: transport/src/notifier.rs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    /// Indicates whether state has succeeded.
+    pub(crate) succeeded: Option<bool>,
+
+    /// The wakers associated with State.
+    pub(crate) wakers: Vec<std::task::Waker>,
+}
+
+/// Notifier allow to wait for a result.
+/// It does not provide timeout mechanism for platform compatibility.
+/// It is recommended to use tokio::time::timeout instead in native-webrtc feature.
```
Comment (RyanKung) [inline transport/src/notifier.rs:22]: `chrono` not work here?

File: transport/src/connections/web_sys_webrtc/mod.rs [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+use crate::ice_server::IceCredentialType;
+use crate::ice_server::IceServer;
+use crate::notifier::Notifier;
+use crate::Transport;
+
+#[derive(Clone)]
+pub struct WebSysWebrtcConnection {
+    webrtc_conn: Arc<RtcPeerConnection>,
+    webrtc_data_channel: Arc<RtcDataChannel>,
+    webrtc_data_channel_open_notifier: Notifier,
```
Comment (RyanKung) [inline transport/src/connections/web_sys_webrtc/mod.rs:41]: It's wired, if connection holding the status promise.

**Review by RyanKung — APPROVED** [review]

**Comment by RyanKung** [comment]
Plz provide a working CURL based handshake example, ref: https://app.gitbook.com/o/QeE6Hayp2irvmutjwDcl/s/lAg1MHUyRI0jsTKcEz8R/advanced-topic/exchange-sdp

I tried once, it's not work (after turn off signature verifier of endpoint

```
➜  rings-node git:(dev) ✗ curl -X POST \
-H "Content-Type: application/json" -H "X-SIGNATURE: foo" \
--data '{"jsonrpc": "2.0", "id": 1, "method": "acceptAnswer", "params": [“some xxxxxxxx”]}’ \
"http://127.0.0.1:50001"
{"jsonrpc":"2.0","error":{"code":209,"message":"Accept answer info failed: Message invalid: Should be ConnectNodeReport."},"id":1}%
```

**Comment by RyanKung** [comment]
```
    Finished dev [unoptimized + debuginfo] target(s) in 0.26s
     Running `target/debug/rings connect node 'http://127.0.0.1:50001'`
Error: Rpc error: Server returned rpc error Server error: Create offer info failed: Transport error: Connection b6072188ec0ceb013eea95f1e5036d5be5096870 already exists.
```

Pending connection should be overwrite, if offer was created twice.

**Comment by Ma233** [comment]
@RyanKung Hi, I updated the handshake operation to close non-connected Connection through the side effect of `get_and_check_connection`.

**Review by RyanKung — APPROVED** [review]

**Pull request merged**